jupyter/jupyter.go

```diff
@@ -299,6 +293,34 @@ func (s *server) execute(c *conn, req *message) error {
 		return s.publishIO("error", content, req)
 	}
 
+	r, w, err := os.Pipe()
+	if err != nil {
+		return errResp(err)
+	}
+	stdout := os.Stdout
+	os.Stdout = w
+
+	buf := new(bytes.Buffer)
+	session.Stdout = buf
+	session.Stderr = buf
+
+	done := make(chan struct{})
+	go func() {
+		io.Copy(buf, r)
+		close(done)
+	}()
+	defer func() {
+		w.Close()
```

Comment:
The timing is off on this. You close the pipe here (when the function exits), and when <-done returns it is safe to read from `buf`. But `buf.String()` is called long before the defer as an argument to publishIO.

I would prefer this code was inside the Exec method in `ngcore`.

(Long-term I think it should go a little deeper, and we should feed the evaluator different stdout/stderr global variables per session, but this pipe trick in ngcore is a good start.)

Reply:
Ah, I see.  Thanks for the look.